### PR TITLE
[libc] fix -Wtype-limits in wctob

### PR DESCRIPTION
**Pull request by nickdesaulniers**
GCC warns:
```
libc/src/__support/wctype_utils.h:33:20: error: comparison of unsigned
expression in ‘< 0’ is always false [-Werror=type-limits]
   33 |   if (c > 127 || c < 0)
      |                  ~~^~~
```

Looking into the signedness of wint_t, it looks like depending on the platform,
__WINT_TYPE__ is defined to int or unsigned int depending on the platform.

Link: https://lab.llvm.org/buildbot/#/builders/250/builds/14891/steps/6/logs/stdio

### CODE REVIEW

File: libc/src/__support/wctype_utils.h [hunk 1]
```diff
@@ -29,9 +29,14 @@ LIBC_INLINE cpp::optional<int> wctob(wint_t c) {
   // This needs to be translated to EOF at the callsite. This is to avoid
   // including stdio.h in this file.
   // The standard states that wint_t may either be an alias of wchar_t or
-  // an alias of an integer type, so we need to keep the c < 0 check.
+  // an alias of an integer type where different platforms define this type with
+  // different signedness, so we need to keep the c < 0 check, hence the
+  // pragmas.
+#pragma GCC diagnostic push
+#pragma GCC diagnostic ignored "-Wtype-limits"
   if (c > 127 || c < 0)
```
Comment (SchrodingerZhu) [inline libc/src/__support/wctype_utils.h:37]: Would `(c & ~127) != 0` be better? (maybe not straightforward enough)

**Comment by nickdesaulniers** [comment]
my SFINAE skills are rusty, but I wonder if `std::is_signed` can be used here?

**Comment by jhuber6** [comment]
> my SFINAE skills are rusty, but I wonder if `std::is_signed` can be used here?

I was just about to comment the same thing. We should be able to do something like this
```c++
if constexpr (cpp::is_signed<wint_t>) {
} else {
}
```

File: libc/src/__support/wctype_utils.h (hunk 1 above)
Comment (michaelrj-google) [reply libc/src/__support/wctype_utils.h:37]: given how simple the condition is, it's cleaner often faster to let the compiler do that sort of optimization. In this specific case, they actually both get compiled to the same result: https://godbolt.org/z/f1E8beTch

File: libc/src/__support/wctype_utils.h [hunk 2]
```diff
@@ -29,9 +29,14 @@ LIBC_INLINE cpp::optional<int> wctob(wint_t c) {
   // This needs to be translated to EOF at the callsite. This is to avoid
   // including stdio.h in this file.
   // The standard states that wint_t may either be an alias of wchar_t or
-  // an alias of an integer type, so we need to keep the c < 0 check.
+  // an alias of an integer type where different platforms define this type with
```
Comment (michaelrj-google) [inline libc/src/__support/wctype_utils.h:32]: nit: remove "where", put a period between "type" and "different".

**Review by michaelrj-google — APPROVED** [review]
LGTM with nit

**Comment by SchrodingerZhu** [comment]
@michaelrj-google sure! I only want to use that to remove the signedness warning :D

**Comment by nickdesaulniers** [comment]
> > my SFINAE skills are rusty, but I wonder if `std::is_signed` can be used here?
> 
> I was just about to comment the same thing. We should be able to do something like this
> 
> ```c++
> if constexpr (cpp::is_signed<wint_t>) {
> } else {
> }
> ```

cool! Done in https://github.com/llvm/llvm-project/pull/74511/commits/f3a22d435ec559ca43b3bd55fb3be10c6eb6e071

File: libc/src/__support/wctype_utils.h (hunk 2 above)
Comment (nickdesaulniers) [reply libc/src/__support/wctype_utils.h]: done in https://github.com/llvm/llvm-project/pull/74511/commits/f3a22d435ec559ca43b3bd55fb3be10c6eb6e071

File: libc/src/__support/wctype_utils.h (hunk 1 above)
Comment (lntue) [reply libc/src/__support/wctype_utils.h]: I think the point is to remove the warnings.  And with that, I think `(c & ~127) != 0` is better than wrapping with `#pragma`'s.  Maybe adding comments to the condition like
```
  This is equivalent to check `(c > 127) || (c < 0)`, but also works without warnings when `wint_t` is unsigned.
```
to make it clearer.

File: libc/src/__support/wctype_utils.h [hunk 3]
```diff
@@ -28,10 +28,14 @@ namespace internal {
 LIBC_INLINE cpp::optional<int> wctob(wint_t c) {
   // This needs to be translated to EOF at the callsite. This is to avoid
   // including stdio.h in this file.
-  // The standard states that wint_t may either be an alias of wchar_t or
-  // an alias of an integer type, so we need to keep the c < 0 check.
-  if (c > 127 || c < 0)
+  if (c > 127)
```
[the remark this answers is not in the document]
Comment (SchrodingerZhu) [reply libc/src/__support/wctype_utils.h:31]: but `wint_t` should always be fine?

File: libc/src/__support/wctype_utils.h [hunk 4]
```diff
@@ -28,15 +28,14 @@ namespace internal {
...
-#pragma GCC diagnostic ignored "-Wtype-limits"
-  if (c > 127 || c < 0)
+  if (c > 127)
     return cpp::nullopt;
-#pragma GCC diagnostic pop
+  // The standard states that wint_t may either be an alias of wchar_t or
+  // an alias of an integer type, different platforms define this type with
+  // different signedness, so we may need the c < 0 check.
+  if constexpr (cpp::is_signed_v<wint_t>)
+    if (c < 0)
```
Comment (lntue) [inline libc/src/__support/wctype_utils.h:37]: Look like this still generates warnings in gcc: https://godbolt.org/z/fTrT3hPx4

Comment (nickdesaulniers) [reply libc/src/__support/wctype_utils.h:38]: done in https://github.com/llvm/llvm-project/pull/74511/commits/0617a61e74faaedc04a3bc7c411e9fa6e8f8cfcd

Comment (lntue) [reply libc/src/__support/wctype_utils.h:37]: Templated version generates warnings for clang instead https://godbolt.org/z/e5qPz6hx9

Comment (SchrodingerZhu) [reply libc/src/__support/wctype_utils.h:37]: How about just `(c & ~127) != 0` with comments then?

Comment (nickdesaulniers) [reply libc/src/__support/wctype_utils.h]: done in https://github.com/llvm/llvm-project/pull/74511/commits/44406f87f465fb2ea5d292801b50a18c9fabca75

**Review by michaelrj-google — APPROVED** [review]
This version is fine, though we could also static cast to an unsigned type and just check if the resulting value is less than 127. Either is fine, feel free to land.

**Review by lntue — APPROVED** [review]

**Pull request merged**